drivers/usb/device/usb_dc_rpi_pico.c

```diff
@@ -625,26 +733,85 @@ int usb_dc_ep_read_wait(uint8_t ep, uint8_t *data,
+ *
+ * @return error code
+ */
+static int usb_dc_control_ep_read_continue(const struct udc_rpi_ep_state *const ep_state,
+	bool *const arm_out_endpoint)
 {
-	struct udc_rpi_ep_state *ep_state = udc_rpi_get_ep_state(ep);
-	size_t bytes_received;
+	const struct usb_setup_packet *const setup = (const void *)&usb_dpram->setup_packet;
+
+	if (state.setup_available) {
+		LOG_DBG("EP0 setup (wLength=%d, is_to_device=%d)",
+			setup->wLength, usb_reqtype_is_to_device(setup));
+		if (setup->wLength != 0U) {
+			/* In the case of a control transfer, we want to prime the OUT endpoint
+			 * exactly once, to either:
+			 * 1) in the to_device case, to receive the data (only if wLength is not 0)
+			 * 2) in the to_host case, to receive a 0-length status stage transfer
+			 *    (only valid if wLength is not 0)
+			 * Note that when wLength = 0, the status stage transfer is always an IN
+			 * type so we don't need to consider that case.
+			 */
+			*arm_out_endpoint = true;
+			state.control_out_ep_rcvd = 0;
+		}
+		state.setup_available = false;
+	} else {
+		const size_t bytes_received = udc_rpi_get_ep_buffer_len(USB_CONTROL_EP_OUT);
+
+		LOG_DBG("EP0_OUT received %d offset: %d", bytes_received, ep_state->read_offset);
+		if (usb_reqtype_is_to_device(setup)) {
+			if (state.control_out_ep_rcvd + ep_state->read_offset < setup->wLength) {
+				/* If no more data in the buffer, but we're still waiting
+				 * for more, start a new read transaction.
+				 */
+				if (bytes_received == ep_state->read_offset) {
+					state.control_out_ep_rcvd += ep_state->read_offset;
+					*arm_out_endpoint = true;
+				}
+			}
+		}
+	}
+	return 0;
+}
+
+int usb_dc_ep_read_continue(const uint8_t ep)
+{
+	const struct udc_rpi_ep_state *const ep_state = udc_rpi_get_ep_state(ep);
+	bool arm_out_endpoint = false;
 
 	if (!ep_state || !USB_EP_DIR_IS_OUT(ep)) {
 		LOG_ERR("Not valid endpoint: %02x", ep);
 		return -EINVAL;
 	}
+	if (ep == USB_CONTROL_EP_OUT) {
+		int ret = usb_dc_control_ep_read_continue(ep_state, &arm_out_endpoint);
 
-	bytes_received = state.setup_available ?
-			 sizeof(struct usb_setup_packet) :
-			 udc_rpi_get_ep_buffer_len(ep);
+		if (ret != 0) {
+			return ret;
+		}
+	} else {
+		const size_t bytes_received = udc_rpi_get_ep_buffer_len(ep);
 
-	state.setup_available = false;
+		LOG_DBG("EP%d_OUT received %d offset: %d", USB_EP_GET_IDX(ep),
+			bytes_received, ep_state->read_offset);
+		/* If no more data in the buffer, start a new read transaction. */
+		if (bytes_received == ep_state->read_offset) {
+			arm_out_endpoint = true;
+		}
+	}
 
-	/* If no more data in the buffer, start a new read transaction. */
-	LOG_DBG("received %d offset: %d", bytes_received, ep_state->read_offset);
-	if (bytes_received == ep_state->read_offset) {
+	if (arm_out_endpoint) {
+		LOG_DBG("Arming EP%d_OUT", USB_EP_GET_IDX(ep));
 		return usb_dc_ep_start_read(ep, DATA_BUFFER_SIZE);
+	} else {
+		LOG_DBG("Not arming EP%d_OUT", USB_EP_GET_IDX(ep));
```

Comment:
`LOG_DBG("Not arming endpoint 0x%02x", ep);`

Reply:
DONE